src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  return actions.reduce((acc, action, index) => {
+    const statePrevious = index === 0 ? { ...state } : { ...acc[index - 1] };
+
+    switch (action.type) {
+      case 'addProperties':
+        acc.push(Object.assign({}, statePrevious, action.extraData));
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete statePrevious[key];
+        }
+        acc.push(Object.assign({}, statePrevious));
```

Comment:
You don't need to use `Object.assign` here since `statePrevious` is already a new object. You can push `statePrevious` directly to the accumulator.